src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -0,0 +1,298 @@
+            .addLast(new MessageNotifier[M](messageSubject))
+        }
+      })
+
+    def initialize: Task[ClientChannelImpl[M]] = {
+      toTask(bootstrap.connect(inetSocketAddress)).map(_ => this)
+    }
+
+    override def sendMessage(message: M): Task[Unit] = {
+
+      Task
+        .fromFuture(activationF)
+        .map(ctx => {
+          log.debug(
+            s"Processing outbound message from local address ${ctx.channel().localAddress()} " +
+              s"to remote address ${ctx.channel().remoteAddress()} via channel id ${ctx.channel().id()}"
+          )
+
+          ctx.writeAndFlush(Unpooled.wrappedBuffer(codec.encode(message)))
+        })
+        .map(_ => ())
+    }
+
+    override def in: Observable[M] = messageSubject
+
+    override def close(): Task[Unit] = {
+      messageSubject.onComplete()
+      Task
+        .fromFuture(activationF)
+        .flatMap(ctx => toTask(ctx.close()))
+        .flatMap(_ => Task.fromFuture(deactivationF))
+    }
+  }
+
+  private val channelSubject = PublishSubject[Channel[InetMultiAddress, M]]()
+
+  private val workerGroup = new NioEventLoopGroup()
+
+  private val clientBootstrap = new Bootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioSocketChannel])
+    .option[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
+    .option[RecvByteBufAllocator](ChannelOption.RCVBUF_ALLOCATOR, new DefaultMaxBytesRecvByteBufAllocator)
+    .option[Integer](ChannelOption.CONNECT_TIMEOUT_MILLIS, 10000)
+
+  private val serverBootstrap = new ServerBootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioServerSocketChannel])
+    .childHandler(new ChannelInitializer[SocketChannel]() {
+      override def initChannel(ch: SocketChannel): Unit = {
+        val newChannel = new ServerChannelImpl[M](ch)
+        channelSubject.onNext(newChannel)
+        log.debug(s"$processAddress received inbound from ${ch.remoteAddress()}.")
+      }
+    })
+    .option[Integer](ChannelOption.SO_BACKLOG, 128)
+    .option[RecvByteBufAllocator](ChannelOption.RCVBUF_ALLOCATOR, new DefaultMaxBytesRecvByteBufAllocator)
+    .childOption[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
+
+  private val serverBind: ChannelFuture = serverBootstrap.bind(config.bindAddress)
+
+  override def initialize(): Task[Unit] =
+    toTask(serverBind).map(_ => log.info(s"Server bound to address ${config.bindAddress}"))
+
+  override def processAddress: InetMultiAddress = config.processAddress
+
+  override def client(to: InetMultiAddress): Task[Channel[InetMultiAddress, M]] = {
+    new ClientChannelImpl[M](to.inetSocketAddress, clientBootstrap).initialize
+  }
+
+  override def server(): Observable[Channel[InetMultiAddress, M]] = channelSubject
+
+  override def shutdown(): Task[Unit] = {
+    channelSubject.onComplete()
+    for {
+      _ <- toTask(serverBind.channel().close())
+      _ <- toTask(workerGroup.shutdownGracefully())
+    } yield ()
+  }
+}
+
+object TLSPeerGroup {
+
+  case class Config(
```

Comment:
Please can you make this identical to the config in DTLSPeerGroup? Either by copying/pasting the config from there or by refactoring it into some common location.